### ISSUE TITLE
BUG: Upload client wipes selected staging area

### ISSUE THREAD

**Issue by daniwelter** [HumanCellAtlas/upload-service#177]
The bug occurred under the following circumstances:
- logged into the EC2 instance with one user multiple times (multiple command line tabs)
- select staging area once
- verified correct staging area selected in each tab
- file upload via upload client with different files in each tab using
`for f in `cat filesaa`; do hca upload files $f; done`
where filesaa/b/c etc is a file with the names of a subset of all the files to be uploaded.

What is observed:
```
Completed 29 MiB/29 MiB with 1 of 1 files remaining
Completed 29 MiB/29 MiB with 1 of 1 files remaining
Completed 29 MiB/29 MiB with 1 of 1 files remaining
Download complete of 21935_7#129_1.fastq.gz to upload area 8dfaa64c-d700-4021-ba03-140353f236f0/21935_7#129_1.fastq.gz
Completed upload of 1 files to upload area 8dfaa64c-d700-4021-ba03-140353f236f0
There is no upload area selected.
Please select one with "/usr/local/bin/hca upload select <uri_or_alias>"
There is no upload area selected.
Please select one with "/usr/local/bin/hca upload select <uri_or_alias>"
```

No external interference with any of the terminal windows, or other activity under the same user (eg accidentally selected a different upload area) occurred between the last successful file upload and the first failed upload due to lack of staging area.

Resetting the correct staging area in one tab restores normal upload processes in the other tabs as well.

**Comment by sampierson** [comment]
To be clear, the upload area itself is not being wiped, but the HCA CLI's is "forgetting" what the current area is.

@daniwelter The current area is stored in `~/.config/hca/config.json`.  If you ever see this again, please run this command: `jq . ~/.config/hca/config.json | grep current` and let us know what it says.

Also, see if the area is still in the list of areas that the hca cli knows about with e.g.: `hca upload areas | grep 8dfaa64c`.

**Comment by sampierson** [comment]
@parthshahva aside: there's a weird message in there "Download complete of ... to upload area".  We're uploading not downloading right?

**Comment by daniwelter** [comment]
@sampierson It always says that at the end of every upload I've ever done.

**Comment by sampierson** [comment]
@parthshahva speculation: Andrey's `tweak.Config` code that `UploadConfig` is built on top of has a `save_on_exit` feature that defaults to `True`.  I bet that running multiple instances of the HCA CLI is causing them to write over each other.

**Comment by parthshahva** [comment]
1) Yup thats a log statement bug. I'll change "Download" to say "Upload"

2) Ah good catch/possibility. I've never seen this happen when only running one hca upload process

**Comment by parthshahva** [comment]
Fixed to upload in https://github.com/HumanCellAtlas/dcp-cli/pull/186

**Comment by diekhans** [comment]
I believe the root issue here is that the hca command saves user-scoped state with "select" subcommand.  This is something a command line utility should not do if it is to be well behaved in a UNIX programming environment.   It will be very frustrating to use this to develop other scripts or even interactively with multiple screens.

The expected way for a command like this to work is to take the input on the command line rather than saving state.

It's not obvious what the goal of "hca select:" is, however, I feel this should be dropped and the URL of the upload area be explicitly provider with each command.

**Comment by diekhans** [comment]
Hi @sampierson

A common approach CLI users would use for work with a repetitive, long argument like the data area URL is to put it in a local variable. Similarly, an opaque token could be set in a variable.   

Have a command modifies the environment the behavior of the command in other shells or invocations of the command in the current shell is confusing and limiting, especially when it isn't documented.  If one were going to try to write scripts with this, it would require creating a lock file so only one set of commands can be run at once.  A goal of UNIX command line is to be able to script other programs with it.

I guess I was not clear about my description of a well behaved UNIX tool.  User-global configuration in home directories is common and understood that changes in the configuration file will change the configuration for new instances of the program.   This is different than a command making dynamic changes to the command being run in other windows.

The shell built-in "cd" only changes the current shell process and its children.  It would be disastrous if cd changed all process that a user had, which is the equivalent to what 'hca select' is doing.  This is very confusing to most people that know UNIX.

Having an explicit argument is good, especially for scripting.  If users want to have a process local environment, a way this is sometimes done is have the command start its own shell.  The cpan command comes to mind.

I suspect we could come up with other ways to do this that are safe and not surprising.  Naming a session on the command line that gets associated with the opaque token comes to mind.

This has already caused confusion and problems with multiple wranglers, so it is a real issue.

**Comment by sampierson** [comment]
@diekhans 
Of course the shell builtin "cd" does not change the environment for all processes a user had.  That would be ridiculous.  Did you seriously think I was trying to imply that?  I added that to navigate you to your home directory before running `ls`.  Let me rephrase that is a less ambiguous fashion for you: `ls -a ~` or `ls -a $HOME`.  FYI I've been using unix systems since the early 80s and was a kernel developer for a time.  You can assume I know how unix systems work.

The point I was making is that in the long term most of our users will **not** be *people that know unix*.  For example it's entirely likely there will be Windows clients as well, and we're trying to design a UX for less-technical people.

However I think we can make all constituencies happy with the addition of a command argument to explicitly specify an upload area for those who wish to, and there is already a shorthand way to refer to upload sessions.

**Comment by diekhans** [comment]
Sorry @sampierson, I was not trying to imply anything about your background (I owe you a beer). I was trying to come up with an analogy that better match the current behavior than configuration files.  Changing directories is the closest thing I can think of for the current behavior.

An argument will solve the cases for people who want to write shell scripts with the hca command.

I believe there are other ways to achieve the goal of a simplified command line user interface than global user state.  It doesn't take much for a user to realize that it is possible to open two windows and start multiple slow tasks.  One then has to document and try to prevent this.

Doing something that is analogous to your  `cd` example would be to create a hca shell.  When you are in the shell, would have the upload area that would are used with defined.  An easy approach that I would be inclined to take would be:

  `hca shell uploadarea`

This would validate uploadarea, get a token and stuff in an environment variable.  It would then launch a child instance of the user's shell with a special prompt, telling them to `exit' when done.

Other invocations of the `hca` command would first look for the token on the command line, if not found, then look in the environment variable.  If not found there, generate an appropriate error.

This gives the current state functionality without sharing it between windows or across login sessions.  Since it is built on the user's shell, it doesn't require much code or documentation.
Personally, I think it  is easier than trying to prevent users from running in multiple windows

There could be other approaches as well. This would be what I would do to achieve your goal.

Mark

**Comment by parthshahva** [comment]
@daniwelter Since each wrangler has a separate user, we don't believe this is a bug at the moment. Closing this ticket with that in mind
